module/mempool/stdmap/eject.go

```diff
@@ -5,50 +5,111 @@ package stdmap
+//    element it removes from the mempool.
+//  * Implementations do _not_ need to be concurrency safe. The Backend handles
+//    concurrency (specifically, it locks the mempool during ejection).
+//  * The implementation should be non-blocking (though, it is allowed to
+//    take a bit of time; the mempool will just be locked during this time).
+type BatchEjectFunc func(b *Backend) bool
+type EjectFunc func(b *Backend) (flow.Identifier, flow.Entity, bool)
 
 // EjectTrueRandom relies on a random generator to pick a random entity to eject from the
 // entity set. It will, on average, iterate through half the entities of the set. However,
 // it provides us with a truly evenly distributed random selection.
-func EjectTrueRandom(entities map[flow.Identifier]flow.Entity) (flow.Identifier, flow.Entity) {
+func EjectTrueRandom(b *Backend) (flow.Identifier, flow.Entity, bool) {
 	var entityID flow.Identifier
 	var entity flow.Entity
+	var bFound bool = false
 	i := 0
-	n := rand.Intn(len(entities))
-	for entityID, entity = range entities {
+	n := rand.Intn(len(b.entities))
+	for entityID, entity = range b.entities {
 		if i == n {
+			bFound = true
 			break
 		}
 		i++
 	}
-	return entityID, entity
+	return entityID, entity, bFound
+}
+
+// EjectTrueRandomFast checks if the map size is beyond the
+// threshold size, and will iterate through them and eject unneeded
+// entries if that is the case.  Return values are unused
+func EjectTrueRandomFast(b *Backend) bool {
+	currentSize := len(b.entities)
+
+	if b.guaranteedCapacity >= uint(currentSize) {
+		return false
+	}
+	// At this point, we know that currentSize > b.guaranteedCapacity. As
+	// currentSize fits into an int, b.guaranteedCapacity must also fit.
+	overcapacity := currentSize - int(b.guaranteedCapacity)
+	if overcapacity <= overCapacityThreshold {
+		return false
+	}
+
+	// Randomly select indices of elements to remove:
+	mapIndices := make([]int, 0, overcapacity)
+	for i := overcapacity; i > 0; i-- {
+		mapIndices = append(mapIndices, rand.Intn(currentSize))
+	}
+	sort.Ints(mapIndices) // inplace
+
+	// Now, mapIndices is a sequentially sorted list of indices to remove.
+	// Remove them in a loop. Repeated indices are idempotent (subsequent
+	// ejection calls will make up for it).
+	idx := 0                     // index into mapIndices
+	next2Remove := mapIndices[0] // index of the element to be removed next
+	i := 0                       // index into the entities map
+	for entityID, entity := range b.entities {
+		if i == next2Remove {
+			delete(b.entities, entityID) // remove entity
+			for _, callback := range b.ejectionCallbacks {
+				callback(entity) // notify callback
+			}
+
+			idx++
+
+			// There is a (1 in b.guaranteedCapacity) chance that the
+			// next value in mapIndices is a duplicate. If a duplicate is
+			// found, skip it by incrementing 'idx'
+			for ; next2Remove == mapIndices[idx] && idx < overcapacity; idx++ {
```

Comment:
⚠️ 🕷️ ❗ 

this code will panic with `index out of range` in the edge case that the index of the _last_ entity was sampled (i.e. `mapIndices` contains `currentSize-1`). 

Note that my suggestion was subtly different:
https://github.com/onflow/flow-go/blob/e0ec08eb66b21fc9f5b38db8d48747d43a1f5234/module/mempool/stdmap/eject.go#L87-L93
* We _always_ enter the loop body, where the index-range check (`idx == overcapacity`) is executed.  _Only_ if we are still within the range, we proceed to evaluate the terminal condition.

```suggestion
			for ; idx < overcapacity && next2Remove == mapIndices[idx]; idx++ {
```